stackslib/src/core/mempool.rs

```diff
@@ -1835,87 +1855,89 @@ impl MemPoolDB {
-                                    expected_sponsor_nonce + 1,
-                                    self.conn(),
-                                );
-                                if !stored {
-                                    Self::save_nonce_for_retry(
-                                        &mut retry_store,
-                                        settings.nonce_cache_size,
+                                if consider.tx.tx.auth.is_sponsored() {
+                                    nonce_cache.set(
                                         consider.tx.metadata.sponsor_address,
                                         expected_sponsor_nonce + 1,
+                                        &mut nonce_conn,
                                     );
                                 }
+                                output_events.push(tx_event);
+                            }
+                            TransactionEvent::Skipped(_) => {
+                                // don't push `Skipped` events to the observer
+                            }
+                            _ => {
+                                output_events.push(tx_event);
                             }
-                            output_events.push(tx_event);
-                        }
-                        TransactionEvent::Skipped(_) => {
-                            // don't push `Skipped` events to the observer
-                        }
-                        _ => {
-                            output_events.push(tx_event);
                         }
                     }
+                    None => {
+                        debug!("Mempool: early exit from iterator");
+                        break MempoolIterationStopReason::IteratorExited;
+                    }
                 }
-                None => {
-                    debug!("Mempool iteration early exit from iterator");
-                    break MempoolIterationStopReason::IteratorExited;
+
+                if settings.strategy == MemPoolWalkStrategy::GlobalFeeRate {
+                    // Reset for finding the next transaction to process
+                    debug!(
+                        "Mempool: reset: retry list has {} entries",
+                        candidate_cache.len()
+                    );
+                    candidate_cache.reset();
+                }
+            };
+
+            // If we've reached the end of the mempool, or if we've stopped
+            // iterating for some other reason, break out of the loop
+            if settings.strategy != MemPoolWalkStrategy::NextNonceWithHighestFeeRate
+                || stop_reason != MempoolIterationStopReason::NoMoreCandidates
+                || !state_changed
+            {
+                if stop_reason == MempoolIterationStopReason::NoMoreCandidates {
+                    debug!("Mempool: no more transactions to consider");
                 }
+                break stop_reason;
             }
```

Comment:
This line isn't clear to my why it's in here, would you mind adding a comment?

```rust
if settings.strategy != MemPoolWalkStrategy::NextNonceWithHighestFeeRate
```

Reply:
Add in 30a68d8d1c. Let me know if that helps or if I should explain something else here.